File: Marriot/DatabaseSystems/02-Document-Database-MongoDB/mongodb-mastery-lab/EXAMPLES/python/aggregation.py
import os
from pprint import pprint

from pymongo import MongoClient

MONGODB_URI = os.getenv(
    "MONGODB_URI",
    "mongodb://app:app_password@localhost:27017/mongodb_mastery?authSource=mongodb_mastery&replicaSet=rs0&directConnection=true",
)

client = MongoClient(MONGODB_URI, serverSelectionTimeoutMS=5000)
db = client.mongodb_mastery

pipeline = [
    {"$match": {"tenantId": "t1", "status": {"$in": ["PAID", "SHIPPED"]}}},
    {"$unwind": "$items"},
    {
        "$group": {
            "_id": "$items.category",
            "units": {"$sum": "$items.quantity"},
            "revenueCents": {
                "$sum": {"$multiply": ["$items.quantity", "$items.priceCents"]}
            },
        }
    },
    {"$sort": {"revenueCents": -1}},
]

for row in db.orders.aggregate(pipeline):
    pprint(row)

client.close()

File: Marriot/DatabaseSystems/02-Document-Database-MongoDB/mongodb-mastery-lab/EXAMPLES/nodejs/src/aggregation.js
import { closeDb, getDb } from './db.js';

async function main() {
  const db = await getDb();
  const orders = db.collection('orders');

  const report = await orders.aggregate([
    { $match: { tenantId: 't1', status: { $in: ['PAID', 'SHIPPED'] } } },
    { $unwind: '$items' },
    {
      $group: {
        _id: '$items.category',
        units: { $sum: '$items.quantity' },
        revenueCents: { $sum: { $multiply: ['$items.quantity', '$items.priceCents'] } }
      }
    },
    { $sort: { revenueCents: -1 } }
  ]).toArray();

  console.table(report.map((row) => ({ category: row._id, units: row.units, revenueCents: row.revenueCents })));
}

main()
  .catch((error) => {
    console.error(error);
    process.exitCode = 1;
  })
  .finally(closeDb);
